addons/sale_coupon/models/sale_order.py

```diff
@@ -141,20 +141,52 @@ def _get_reward_values_discount_percentage_per_line(self, program, line):
         discount_amount = line.product_uom_qty * line.price_reduce * (program.discount_percentage / 100)
         return discount_amount
 
+    def _get_reward_values_fixed_amount(self, program):
+
+        discount_amount = - self._get_reward_values_discount_fixed_amount(program)
+
+        lines = self._get_paid_order_lines()
+        # Remove Free Lines
+        lines = lines.filtered('price_reduce')
+        lines_total_amount = sum(lines.filtered(lambda l: l.price_total > 0).mapped('price_total'))
+        reward_lines = {}
+
+        for line in lines:
+
+            taxes = line.tax_id
+
+            line_price = line.price_reduce * line.product_uom_qty
+            # Next line is equivalent to computing the % of taxes included price then
+            # getting the related sale_price
+            # discount_line_amount_tax_included =
+            #   discount_amount * line.price_total / lines_total_amount
+            # discount_line_amount_price =
+            #   discount_line_amount_tax_included * line_price / line.price_total
+            discount_line_amount_price = discount_amount * line_price / lines_total_amount
+
+            if taxes in reward_lines:
+                reward_lines[taxes]['price_unit'] += discount_line_amount_price
+            else:
+                tax_name = ""
+                if len(taxes) == 1:
+                    tax_name = " - " + _("On product with following tax: ") + ', '.join(taxes.mapped('name'))
+                elif len(taxes) > 1:
+                    tax_name = " - " + _("On product with following taxes: ") + ', '.join(taxes.mapped('name'))
+
+                reward_lines[taxes] = {
+                    'name': _("Discount: ") + program.name + tax_name,
+                    'product_id': program.discount_line_product_id.id,
+                    'price_unit': discount_line_amount_price,
+                    'product_uom_qty': 1.0,
+                    'product_uom': program.discount_line_product_id.uom_id.id,
+                    'is_reward_line': True,
+                    'tax_id': [(4, tax.id, False) for tax in line.tax_id],
+                }
+        return reward_lines.values()
+
     def _get_reward_values_discount(self, program):
         if program.discount_type == 'fixed_amount':
-            taxes = program.discount_line_product_id.taxes_id
-            if self.fiscal_position_id:
-                taxes = self.fiscal_position_id.map_tax(taxes)
```

Comment:
Just to summarize and confirm.

This is just removed, not adapted. Which seems indeed correct, as it was done for people who added taxes manually on their promotion program product (the one created when you create your promotion).

This is not something rare, probably mostly due to the fact the fixed amount discount lines did not get a tax applied (which you are fixing with this PR).

Eg, https://github.com/odoo-dev/odoo/commit/7633f963066c4962a5d722b7e755e8468b4c2b54

Everything still seems to be working fine for those cases, as the applied tax on that product will just be ignored and the one from the SOL the reward is applied will be considered.
(And that SOL tax is already mapped, so the discount line already received the mapped tax)

Reply:
HEllo, I confirm that I prefer Romain's suggestion.
Also, if the product generated by the program has a tax, we should use that one.
That way, admin can specify the tax they want on the product.